src/object_store/src/object/s3.rs

```diff
@@ -12,22 +12,259 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::sync::{Arc, Mutex};
+
+use aws_sdk_s3::model::{CompletedMultipartUpload, CompletedPart};
+use aws_sdk_s3::output::UploadPartOutput;
 use aws_sdk_s3::{Client, Endpoint, Region};
 use fail::fail_point;
 use futures::future::try_join_all;
+use futures::stream;
+use hyper::Body;
 use itertools::Itertools;
+use tokio::task::JoinHandle;
+
+use super::object_metrics::ObjectStoreMetrics;
+use crate::object::{
+    BlockLocation, Bytes, ObjectError, ObjectMetadata, ObjectResult, ObjectStore,
+    StreamingUploader, StreamingUploaderImpl,
+};
+
+type PartId = i32;
+
+const MIN_PART_ID: PartId = 1;
+const MIN_PART_SIZE: usize = 5 * 1024 * 1024;
+const PART_SIZE: usize = 16 * 1024 * 1024;
+
+/// S3 multipart upload handle.
+/// Reference: <https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html>
+pub struct S3StreamingUploader {
+    client: Arc<Client>,
+    bucket: String,
+    /// The key of the object.
+    key: String,
+    /// The identifier of multipart upload task for S3.
+    upload_id: String,
+    /// Next part ID.
+    next_part_id: PartId,
+    /// Join handles for part uploads.
+    join_handles: Vec<JoinHandle<ObjectResult<()>>>,
+    /// Parts that are already uploaded to S3.
+    uploaded_parts: Arc<Mutex<Vec<(PartId, UploadPartOutput)>>>,
+    /// Buffer for bytes.
+    buf: Vec<Bytes>,
```

Comment:
It is also worth mentioning the following considerations for buf:
1. why we use`Vec<Bytes>`
2. the fact that we keep two part in the buf and why

Reply:
I will explain them in the comments.